### PR TITLE
17649 - Hardware wallet signing state

### LINKED ISSUE #17606: [Bug]: Vague Error Message when Connecting to Locked Ledger
### Describe the bug

Vague error message for connecting to Ledger devices that are locked.  

> 🔴 NOTE: this might seem trivial but is likely quite confusing for first-time Ledger users.

### Steps to reproduce

1. Ensure that your Ledger is locked (i.e., the Ledger screensaver is showing on the device)
2. Connect to your Ledger without first inserting your pin
image

### Error messages or log output

_No response_

### Version

10.24.2

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

Ledger

### Additional context

_No response_

### LINKED ISSUE #17630: [Bug]: Silent Ledger Failures on Various Popups
### Describe the bug

Issue explained via story:
* User initiates a transaction with an unlocked Ledger
* Confirmation Modal pops up, and no warning is present because the Ledger is unlocked
* Ledger locks while the Confirmation is open (e.g. the user may be hesitating the confirmation for a bit, or may be playing around with fees etc.)
* User hits "Confirm" and the transaction fails silently as the Ledger is now locked
 
Issue is applicable to Confirmation modals for transactions, signing typed data, signing messages etc.

### Steps to reproduce

1. Initiate transaction with Ledger
2. While on the confirm screen, wait for the Ledger to lock
3. Confirm transaction and you'll receive no notice of the confirmation failing

### Error messages or log output

_No response_

### Version

develop

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

Ledger

### Additional context

_No response_

### LINKED ISSUE #17649: Confirmation State for Hardware Wallet Signing
### Describe the bug

**Feature would entail:**
Similar to how there's a loading screen when connecting a wallet, I'm thinking that there could be a similar loading screen while waiting for hardware wallet interaction

**Reasons for the feature:**
* Supports MV3 Hardware wallet changes as we can persist the client until the transaction is finished
* There are some user stories that benefit from a loading screen:
  * E.g., immediately directs user's attention to confirm the transaction/signing on their device. A new hardware wallet user would intuitively think that the transaction is finished when they hit "Confirm", esp. if this user is used to using hard-disk wallets
  * E.g., there are some situations where the app may appear broken if there's a pending transaction to be signed on the wallet - especially relevant to Ledgers

image

### Steps to reproduce

1.

### Error messages or log output

_No response_

### Version

develop

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by flexa-rob**
## Explanation

Issue explained via story:

1. User initiates a signing request with an unlocked Ledger
2. Confirmation Modal pops up, and no warning is present because the Ledger is unlocked
3. Ledger locks while the Confirmation is open (e.g. the user may be hesitating the confirmation for a bit, or may be playing around with fees etc.)
4. User hits "Sign" and the signing request fails silently as the Ledger is now locked

Fixes: 
#17649 
#17630
#17606

See related:
#17937 
#17940

### Before

Signing request was allowed to proceed with a locked device:
image

Resulting in a silent error and no result:
image

### After

Signing requests now monitor HD wallet state and displays an appropriate error message when locked.

Personal Sign:
image

Sign Typed Data:
image

Sign Typed Data V3:
image

Sign Typed Data V4:
image

Sign in With Ethereum:
image

Send Transaction:
image

## Manual Testing Steps

1. Initiate signing request with Ledger
2. While on the confirm screen, wait for the Ledger to lock. Signing request should be blocked with ledgerLocked warning message showing
3. Unlock Ledger. The ledgerLocked warning should no longer be present and the signing request should now be able to proceed
4. Sign request and proceed as normal

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:56]: You should be able to use BannerAlert here. ActionableMessage is in the process of being deprecated. I'll create an issue to deprecate ActionableMessage.

https://metamask.github.io/metamask-storybook/?path=/docs/components-componentlibrary-banneralert--severity

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Still need to do a comprehensive review and check out the branch but one suggestion. Great work on the story creation!

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:56]: Thanks @georgewrmarshall - I'll make that change later today

**Comment by Babayboy48** [comment]
Hu

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (georgewrmarshall) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js]: Thanks @flexa-rob will try re-review today

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:62]: PropTypes descriptions are always nice!

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:53]: I was going to suggest separating the description into a heading and description but fitting it all into the description at a smaller font size works too. Checked storybook build looks good!

Screenshot 2023-02-21 at 10 24 26 PM

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js:6]: nit: remove line break
```suggestion
```

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js:18]: nit: remove line break
```suggestion
```

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js:1]: I would consider this an app specific component. It doesn't have a lot of reusability and is for a very specific purpose. It could probably live in `ui/components/app`. There isn't much documentation on `App` vs `UI` components but the difference is that `UI` components should have very high reusability. The folders are currently pretty muddy in terms of definition. Hopefully we can work on improving that.

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js:12]: Could improve the use of this control. Not that I think it adds a huge amount of value in storybook form but can't hurt.

```suggestion
      control: 'select',
      options: ['locked', 'unlocked'],
```

File: ui/components/ui/hardware-wallet-state/index.scss
Comment (georgewrmarshall) [inline ui/components/ui/hardware-wallet-state/index.scss:5]: Not sure this file is needed anymore?

**Review by georgewrmarshall — COMMENTED** [review]
Looking great! Left a few minor suggestions. Travelling right now so don't have a hardware wallet on me to test locally so will need another approval to test funcitonality. I think it's out of the scope of this ticket but it would be great to update this notification to use BannerAlert as well. Might need some design attention 😅 

Screenshot 2023-02-21 at 10 37 27 PM

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:62]: Done - thanks

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js:53]: yeah, I just based it off the `create-account/connect-hardware` page. The message is relatively long, so already takes up some space in the popup modal - though happy to add a title if you recommend it (just let me know which message to use)

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.stories.js:1]: Done - thanks for the explanation. It started in `App`, then I moved it to `UI` and then I had second thoughts :joy:

File: ui/components/ui/hardware-wallet-state/index.scss
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/index.scss:5]: I went back and forth a few times of needing and not needing it - removed

**Comment by flexa-rob** [comment]
> Looking great! Left a few minor suggestions. Travelling right now so don't have a hardware wallet on me to test locally so will need another approval to test funcitonality. I think it's out of the scope of this ticket but it would be great to update this notification to use BannerAlert as well. Might need some design attention sweat_smile
> 
> Screenshot 2023-02-21 at 10 37 27 PM

Thanks @georgewrmarshall - I just looked into updating the `ledger-instruction-field` too. 

I'd be happy to do this, though would recommend that it be implemented in a separate PR - it uses `Typography` under the covers with some `Button` functionality in certain cases. So, there's probably enough effort (and testing) to warrant a separate PR/issue, if that's ok with you?

File: ui/components/ui/hardware-wallet-state/index.scss
Comment (georgewrmarshall) [reply ui/components/ui/hardware-wallet-state/index.scss]: Oh I wasn't aware it was used. I couldn't find it as a className anywhere? Is it possibly left over from when the old component was used?

Screenshot 2023-02-22 at 5 52 31 PM

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (georgewrmarshall) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js]: Fine like this I think, have you got a design contact for this ticket?

**Comment by georgewrmarshall** [comment]
@flexa-rob absolutely! Created another ticket here https://github.com/MetaMask/metamask-extension/issues/17873

File: ui/components/ui/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/hardware-wallet-state.js]: ok, thanks. Do you mean a contact in the design team? I haven't been given anyone's details.

File: ui/components/ui/hardware-wallet-state/index.scss
Comment (flexa-rob) [reply ui/components/ui/hardware-wallet-state/index.scss]: Sorry, I have removed it - it wasn't used (anymore)

**Review by georgewrmarshall — COMMENTED** [review]
This is looking really really great. I've made some small suggestions. There is one issue when I sign I get the full screen mode and for some reason the signature container is set to a static height. I'm quite tempted to remove it be cause it's not scalable at all. We might need some design feedback and I'm trying to find out if there is any reason why it is set. It could be that other screens have an overflow of scroll and sticky footer for buttons but I'm not sure.

Screenshot 2023-03-02 at 4 21 33 PM

File: ui/components/app/hardware-wallet-state/index.scss
Comment (flexa-rob) [reply ui/components/app/hardware-wallet-state/index.scss:5]: ok, thanks - I'll make these updates

Comment (flexa-rob) [reply ui/components/app/hardware-wallet-state/index.scss]: ok, great idea, done

File: ui/components/app/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/app/hardware-wallet-state/hardware-wallet-state.js]: done, I was able to get rid of `className` and just pass it through also

File: ui/components/app/hardware-wallet-state/hardware-wallet-state.js
Comment (flexa-rob) [reply ui/components/app/hardware-wallet-state/hardware-wallet-state.js]: done

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (flexa-rob) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js]: done - looks good. Thanks

**Comment by georgewrmarshall** [comment]
Looking great @flexa-rob! I think we just remove this static height here I can't see any reason for it? If it's removed does it still open to a reasonable height on your machine? 

https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/signature-request-original/index.scss#L22-L24

**Review by georgewrmarshall — COMMENTED** [review]
Looking great @flexa-rob! I think we just remove this static height here I can't see any reason for it? If it's removed does it still open to a reasonable height on your machine? 

https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/signature-request-original/index.scss#L22-L24

**Comment by flexa-rob** [comment]
> Looking great @flexa-rob! I think we just remove this static height here I can't see any reason for it? If it's removed does it still open to a reasonable height on your machine?
> 
> https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/signature-request-original/index.scss#L22-L24

Thanks @georgewrmarshall  - sure, I'll remove the static height today and we can see how it looks.

**Comment by flexa-rob** [comment]
> This is looking really really great. I've made some small suggestions. There is one issue when I sign I get the full screen mode and for some reason the signature container is set to a static height. I'm quite tempted to remove it be cause it's not scalable at all. We might need some design feedback and I'm trying to find out if there is any reason why it is set. It could be that other screens have an overflow of scroll and sticky footer for buttons but I'm not sure.
> 
> Screenshot 2023-03-02 at 4 21 33 PM

@georgewrmarshall - when I remove the explicit height (from `@include screen-sm-min` on L23) it takes up the whole screen:
image

If I disable `flex-grow` (`flex: 1 0 auto`) the `main-container-wrapper`, then it works fine.. though that seems to be a common setting that could have unexpected side-effects if messed with..

image

**Comment by georgewrmarshall** [comment]
What if we set `height: max-content;`? here https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/signature-request-original/index.scss#L12

**Comment by flexa-rob** [comment]
> What if we remove height (from @include screen-sm-min on L23) and set `height: max-content;` here https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/signature-request-original/index.scss#L12

Nice CSS-fu - looks good: 
image

File: ui/components/app/signature-request-original/index.scss
Comment (georgewrmarshall) [inline ui/components/app/signature-request-original/index.scss:22]: We cam remove this mixin all together it just puts any styles inside a media query so if there are no styles it's not required.
```suggestion
```

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Looking great one tiny suggestion then I will approve

File: ui/components/app/signature-request-original/index.scss
Comment (flexa-rob) [reply ui/components/app/signature-request-original/index.scss]: ok, done - thanks

Let me know if there are any other tickets you want a hand with we're while waiting for another approval

**Review by georgewrmarshall — DISMISSED** [review]

**Review by georgewrmarshall — DISMISSED** [review]

**Review by georgewrmarshall — DISMISSED** [review]

**Review by georgewrmarshall — DISMISSED** [review]

**Review by georgewrmarshall — DISMISSED** [review]

**Comment by bschorchit** [comment]
Do we really want to use the red banner for the ledger information here? 
For context, we're using the same red banner for flagging a malicious transaction to users and warn them not to go forward with it. I think we should preserve it for really dangerous-level information. Besides this making the screen crowded as Mariona pointed out above.

Could we dynamically update the content of the blue ledger box to display this information instead?

**Comment by flexa-rob** [comment]
> Do we really want to use the red banner for the ledger information here? For context, we're using the same red banner for flagging a malicious transaction to users and warn them not to go forward with it. I think we should preserve it for really dangerous-level information. Besides this making the screen crowded as Mariona pointed out above.
> 
> Could we dynamically update the content of the blue ledger box to display this information instead?

Thanks @bschorchit and @seaona - sure, I can integrate the message into the ledger instruction banner. 

Having the device disconnected does prevent the transaction from proceeding (transaction fails with a vague error message). So in that case, it's probably more appropriate to show a pre-emptive warning (orange) or error (red) styling, rather than the regular (blue) informational message.

Perhaps the ledger instruction field should be rendered in orange, along with the "Cannot connect" message, if the device is disconnected?

Also (FYI) - here is a separate PR to bring the ledger instruction field styling into line with the new styling: #17873 . It has one UX/UI approval and requires one more before we can merge.

**Comment by flexa-rob** [comment]
@bschorchit / @seaona , I have integrated the message into the ledger instruction banner, which will render orange (warning) if the HD wallet is locked or disconnected. I also had to update the styling to avoid merge conflicts with #17937 (which is safe to merge once approved) 

Screenshots above have been updated to reflect these changes.

File: ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js:59]: I think we could remove these and just make the default story show the banner? I'm not sure these add too much more value?
```suggestion
```

File: ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (georgewrmarshall) [inline ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js:62]: Update the DefaultStory to show the banner
```suggestion
  <Provider store={customStore({ hardwareWalletState: HardwareWalletStates.locked })}>
```

**Review by georgewrmarshall — COMMENTED** [review]
Hey @flexa-rob, was doing another re-review of the UI and have made some small storybook suggestions. Otherwise this LGTM from a UI perspective. Going to put this out internally to see if I can get another reviewer to check the redux logic.

File: ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js
Comment (flexa-rob) [reply ui/components/app/hardware-wallet-state/hardware-wallet-state.stories.js]: Thanks @georgewrmarshall - yeah, agreed. I've applied these suggestions.

**Review by georgewrmarshall — DISMISSED** [review]
UI LGTM! 💯

**Review by georgewrmarshall — DISMISSED** [review]
Tested on local UI LGTM!


https://github.com/MetaMask/metamask-extension/assets/8112138/80e1d984-02a0-411a-a74b-62faf1ed3f3c

**Review by Silvrbckw — APPROVED** [review]
Commit and push

**Pull request closed without merging**
